util/internal/integrationtest/common/src/main/java/org/hibernate/search/util/impl/integrationtest/common/rule/SearchWorkCall.java

```diff
@@ -92,9 +93,15 @@ protected boolean isSimilarTo(SearchWorkCall<?> other) {
 
 		for ( Object rawHit : rawHits ) {
 			actualProjectionContext.reset();
-			extractedElements.add(
-					actualRootProjection.extract( actualProjectionHitMapper, rawHit, actualProjectionContext )
-			);
+			Object extract = actualRootProjection.extract( actualProjectionHitMapper, rawHit, actualProjectionContext );
+			extractedElements.add( extract );
+
+			// Check for HSEARCH-3896
+			if ( extract instanceof Integer ) {
+				if ( ( (int) extract ) >= rawHits.size() ) {
```

Comment:
I wouldn't add this check. Keys do not necessary match the index of hits.

For example you could have two requested loads for a single hit. That may be the reason this checks fails?

Currently that's a bit dumb, since we would load the same entity twice, but in the future one can imagine loading associated entities from their ID specified in a field, for example.

Reply:
dropped this commit